projects/frontegg-app/src/lib/guards/frontegg-auth.guard.ts

```diff
@@ -14,33 +13,62 @@ export class FronteggAuthGuard extends FronteggBaseGuard {
-    if (this.fronteggAppService.fronteggApp.options.hostedLoginBox) {
-      localStorage.setItem('FRONTEGG_AFTER_AUTH_REDIRECT_URL', redirectUrl);
-      this.fronteggAuthService.loginWithRedirect();
-      return false;
-    } else {
-      return this.router.navigateByUrl(this.fronteggAppService.authRoutes.loginUrl + '?redirectUrl=' + encodeURIComponent(redirectUrl));
+
+  /**
+   * Wait for loader to finish
+   * @private
+   */
+  private waitForLoader(): Promise<boolean> {
+    return new Promise((resolve) => {
+      const subscription = this.fronteggAppService.isLoading$.subscribe((isLoading) => {
+        if (!isLoading) {
+          resolve(true);
+          subscription.unsubscribe();
+        }
+      });
+    });
+  }
+
+  /**
+   * Navigate to login page if user is not authenticated
+   * @private
+   */
+  private async navigateToLoginIfNeeded(redirectUrl: string): Promise<boolean> {
+    const { isAuthenticated } = this.fronteggAuthService.getAuthState();
+    if (!isAuthenticated) {
+      if (this.fronteggAppService.fronteggApp.options.hostedLoginBox) {
+        localStorage.setItem('FRONTEGG_AFTER_AUTH_REDIRECT_URL', redirectUrl);
+        this.fronteggAuthService.loginWithRedirect({ prompt: 'login' });
+        return false;
+      } else {
+        return this.router.navigateByUrl(this.fronteggAppService.authRoutes.loginUrl + '?redirectUrl=' + encodeURIComponent(redirectUrl));
+      }
+      return false; // prevent navigation
     }
+    return true; // activate navigation
   }
 
   canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
     const redirectUrl = state.url;
-    return new Promise<boolean>((resolve, reject) => {
-      const subscription = this.fronteggAppService.isAuthenticated$.pipe(take(1)).subscribe((isAuthenticated: boolean) => {
-        if (isAuthenticated) {
-          resolve(true);
-        } else {
-          if (NgZone.isInAngularZone()) {
-            this.navigateToUrl(redirectUrl).then(resolve, reject);
-          } else {
-            this.ngZone.run(() => {
-              this.navigateToUrl(redirectUrl).then(resolve, reject);
-            });
-          }
-        }
-      });
-      subscription.unsubscribe();
+    const { isLoading } = this.fronteggAuthService.getAuthState();
+
+    if (!isLoading) {
+      // if showLoader false
+      // check if user is authenticated
+      return this.navigateToLoginIfNeeded(redirectUrl);
+    }
+
+    // wait for loader to finish and then check if user is authenticated
+    return new Promise<boolean>(async (resolve, reject) => {
+      await this.waitForLoader();
+
+      if (NgZone.isInAngularZone()) {
+        this.navigateToLoginIfNeeded(redirectUrl).then(resolve).catch(reject);
+      } else {
+        this.ngZone.run(() => {
+          this.navigateToLoginIfNeeded(redirectUrl).then(resolve).catch(reject);
+        });
```

Comment:
Since we have this pattern in multiple places, it can be nice to create a function that can handle this.
something like `runInNgZoneIfNeeded`

Reply:
done